src/trait_group.h

```diff
@@ -0,0 +1,202 @@
+#pragma once
+#ifndef TRAIT_GROUP_H
+#define TRAIT_GROUP_H
+
+#include "json.h"
+#include "mutation.h"
+#include "string.h"
+#include "string_id.h"
+
+typedef std::string Trait_group_tag;
+typedef std::vector<trait_id> Trait_list;
+
+namespace trait_group
+{
+
+/**
+ * Returns a randomized list of traits from the given trait group.
+ */
+Trait_list traits_from( const Trait_group_tag &gid );
+
+/**
+ * Check whether a particular trait is a member of the given group.
+ */
+bool group_contains_trait( const Trait_group_tag &gid, const trait_id &tid );
+
+/**
+ * Check whether the given group ID has been defined.
+ */
+bool group_is_defined( const Trait_group_tag &gid );
+
+/**
+ * See @ref mutation_branch::load_trait_group
+ */
+void load_trait_group( JsonObject &jsobj, const Trait_group_tag &gid, const std::string &subtype );
+
+
+/**
+ * Get a trait group ID and optionally load an inlined trait group.
+ *
+ * If the next value in the JSON stream is string, it's assumed to be a trait group id and it's
+ * returned directly.
+ *
+ * If the next value is a JSON object, it is loaded as a trait group. The group will be given a
+ * unique id (if the JSON object contains an id, it is ignored) and that id will be returned.
+ * If the JSON object does not contain a subtype, the given default is used.
+ *
+ * If the next value is a JSON array, it is loaded as a trait group: the default_subtype will be
+ * used as subtype of the new trait group and the array is loaded like the "entries" array of
+ * a trait group definition (see format of trait groups).
+ *
+ * @param stream Stream to load from
+ * @param default_subtype If an inlined trait group is loaded this is used as the default
+ * subtype. It must be either "distribution" or "collection". See @ref Trait_group.
+ * @throw std::string as usual for JSON errors, including invalid input values.
+ */
+Trait_group_tag load_trait_group( JsonIn &stream, const std::string &default_subtype );
+
+/**
+ * Show a debug menu for testing trait groups.
+ */
+void debug_spawn();
+
+}
+
+/**
+ * Base interface for trait generation.
+ * Used to generate a list of traits.
+ */
+class Trait_creation_data
+{
+    public:
+        typedef std::vector<Trait_group_tag> RecursionList;
+
+        Trait_creation_data( int _probability ) : probability( _probability ) {}
+        virtual ~Trait_creation_data() {};
```

Comment:
You can write ` = default;` instead of the empty function body. It's sometimes more efficient. And there are some semantic differences, generally `default` is not worse than a default, but potentially better.